src/ray/raylet/raylet_client.h

```diff
@@ -0,0 +1,180 @@
+#ifndef RAYLET_CLIENT_H
+#define RAYLET_CLIENT_H
+
+#include <unistd.h>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
+#include "ray/raylet/task_spec.h"
+#include "ray/status.h"
+
+using ray::ActorID;
+using ray::JobID;
+using ray::ObjectID;
+using ray::TaskID;
+using ray::UniqueID;
+
+using MessageType = ray::protocol::MessageType;
+using ResourceMappingType =
+    std::unordered_map<std::string, std::vector<std::pair<int64_t, double>>>;
+
+class RayletConnection {
+ public:
+  /// Connect to the local scheduler.
+  ///
+  /// \param raylet_socket The name of the socket to use to connect to the
+  ///        local scheduler.
+  /// \param worker_id A unique ID to represent the worker.
+  /// \param is_worker Whether this client is a worker. If it is a worker, an
+  ///        additional message will be sent to register as one.
+  /// \param driver_id The ID of the driver. This is non-nil if the client is a
+  ///        driver.
+  /// \return The connection information.
+  RayletConnection(const std::string &raylet_socket, int num_retries, int64_t timeout);
+
+  ~RayletConnection() { close(conn_); }
+  /// Notify the local scheduler that this client is disconnecting gracefully. This
+  /// is used by actors to exit gracefully so that the local scheduler doesn't
+  /// propagate an error message to the driver.
+  ///
+  /// \return ray::Status.
+  ray::Status Disconnect();
+  ray::Status ReadMessage(MessageType type, uint8_t *&message);
+  ray::Status WriteMessage(MessageType type,
+                           flatbuffers::FlatBufferBuilder *fbb = nullptr);
+  ray::Status AtomicRequestReply(MessageType request_type, MessageType reply_type,
+                                 uint8_t *&reply_message,
+                                 flatbuffers::FlatBufferBuilder *fbb = nullptr);
+
+ private:
+  // TODO(rkn): The io methods below should be removed.
+  int connect_ipc_sock(const std::string &socket_pathname);
+  int read_bytes(uint8_t *cursor, size_t length);
+  int write_bytes(uint8_t *cursor, size_t length);
```

Comment:
for these helper functions that doesn't involve class member variables, maybe good to have them as non-member-functions and thus don't need to appear in header file?